lib/rules/no-set-state.js

```diff
@@ -29,22 +44,49 @@ module.exports = {
      * @param {Object} component The component to process
      * @returns {Boolean} True if the component is valid, false if not.
      */
+
+    const configuration = context.options[0] || {};
+    const notAllowedIn = configuration.in;
+
     function isValid(component) {
       return Boolean(component && !component.useSetState);
     }
 
+    function getMethodName(node) {
+      let parent = node.parent;
+      while (parent) {
+        if ((parent.type === 'ClassProperty' && parent.value.type === 'ArrowFunctionExpression') ||
+          (parent.type === 'Property' && parent.value.type === 'FunctionExpression') ||
+          (parent.type === 'MethodDefinition')
+        ) {
+          return parent.key.name;
+        }
+        parent = parent.parent;
+      }
+      return null;
+    }
+
     /**
      * Reports usages of setState for a given component
      * @param {Object} component The component to process
      */
     function reportSetStateUsages(component) {
       let setStateUsage;
-      for (let i = 0, j = component.setStateUsages.length; i < j; i++) {
-        setStateUsage = component.setStateUsages[i];
-        context.report({
-          node: setStateUsage,
-          message: 'Do not use setState'
-        });
+      if (notAllowedIn && notAllowedIn.length && component.setStateUsages.length) {
```

Comment:
let's just delete the else case entirely, and unconditionally use the `if` case.